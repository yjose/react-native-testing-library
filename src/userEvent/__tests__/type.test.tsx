import * as React from 'react';
import { TextInput } from 'react-native';
import { render } from '../..';
import { userEvent } from '..';

interface EventEntry {
  id: number;
  name: string;
  nativeEvent: any;
}

function createEventToolkit() {
  let id = 0;
  const events: EventEntry[] = [];
  const handleEvent = (name: string) => {
    return (event: any) => {
      events.push({
        id,
        name,
        nativeEvent: event?.nativeEvent,
      });

      id += 1;
    };
  };

  return { events, handleEvent };
}

function renderTextInputWithToolkit() {
  const { events, handleEvent } = createEventToolkit();

  const screen = render(
    <TextInput
      testID="input"
      onFocus={handleEvent('focus')}
      onBlur={handleEvent('blur')}
      onPressIn={handleEvent('pressIn')}
      onPressOut={handleEvent('pressOut')}
      onChange={handleEvent('change')}
      onChangeText={handleEvent('changeText')}
      onTextInput={handleEvent('textInput')}
      onKeyPress={handleEvent('keyPress')}
      onSelectionChange={handleEvent('selectionChange')}
      onSubmitEditing={handleEvent('onSubmitEditing')}
      onContentSizeChange={handleEvent('contentSizeChange')}
    />
  );

  return {
    ...screen,
    events,
  };
}

test('userEvent.type on TextInput', async () => {
  const { events, ...queries } = renderTextInputWithToolkit();

  await userEvent.type(queries.getByTestId('input'), 'Hello');

  const entryEvents = events.slice(0, 3);
  const exitEvents = events.slice(events.length - 2);
  const typingEvents = events.slice(
    entryEvents.length,
    events.length - exitEvents.length
  );

  expect(entryEvents).toMatchSnapshot('entry events');
  expect(typingEvents).toMatchSnapshot('typing events');
  expect(exitEvents).toMatchSnapshot('exit events');
});
