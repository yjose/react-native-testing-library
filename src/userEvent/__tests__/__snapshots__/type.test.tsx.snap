// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userEvent.type on TextInput: entry events 1`] = `
[
  {
    "id": 0,
    "name": "pressIn",
    "nativeEvent": undefined,
  },
  {
    "id": 1,
    "name": "focus",
    "nativeEvent": {},
  },
  {
    "id": 2,
    "name": "pressOut",
    "nativeEvent": undefined,
  },
]
`;

exports[`userEvent.type on TextInput: exit events 1`] = `
[
  {
    "id": 28,
    "name": "onSubmitEditing",
    "nativeEvent": undefined,
  },
  {
    "id": 29,
    "name": "blur",
    "nativeEvent": undefined,
  },
]
`;

exports[`userEvent.type on TextInput: typing events 1`] = `
[
  {
    "id": 3,
    "name": "keyPress",
    "nativeEvent": {
      "key": "H",
    },
  },
  {
    "id": 4,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "",
      "text": "H",
    },
  },
  {
    "id": 5,
    "name": "change",
    "nativeEvent": {
      "text": "H",
    },
  },
  {
    "id": 6,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 7,
    "name": "selectionChange",
    "nativeEvent": undefined,
  },
  {
    "id": 8,
    "name": "keyPress",
    "nativeEvent": {
      "key": "e",
    },
  },
  {
    "id": 9,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "H",
      "text": "He",
    },
  },
  {
    "id": 10,
    "name": "change",
    "nativeEvent": {
      "text": "He",
    },
  },
  {
    "id": 11,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 12,
    "name": "selectionChange",
    "nativeEvent": undefined,
  },
  {
    "id": 13,
    "name": "keyPress",
    "nativeEvent": {
      "key": "l",
    },
  },
  {
    "id": 14,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "He",
      "text": "Hel",
    },
  },
  {
    "id": 15,
    "name": "change",
    "nativeEvent": {
      "text": "Hel",
    },
  },
  {
    "id": 16,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 17,
    "name": "selectionChange",
    "nativeEvent": undefined,
  },
  {
    "id": 18,
    "name": "keyPress",
    "nativeEvent": {
      "key": "l",
    },
  },
  {
    "id": 19,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "Hel",
      "text": "Hell",
    },
  },
  {
    "id": 20,
    "name": "change",
    "nativeEvent": {
      "text": "Hell",
    },
  },
  {
    "id": 21,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 22,
    "name": "selectionChange",
    "nativeEvent": undefined,
  },
  {
    "id": 23,
    "name": "keyPress",
    "nativeEvent": {
      "key": "o",
    },
  },
  {
    "id": 24,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "Hell",
      "text": "Hello",
    },
  },
  {
    "id": 25,
    "name": "change",
    "nativeEvent": {
      "text": "Hello",
    },
  },
  {
    "id": 26,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 27,
    "name": "selectionChange",
    "nativeEvent": undefined,
  },
]
`;
